### PR TITLE
Use a different JSON serializer

### PR DESCRIPTION
**Pull request by pablobaxter**
This PR does the following:
- Update several libraries
- Stops using Android SDK variants of an API in favor for the open source variant (ex. `androidx.collection.ArraySet` vs `android.util.ArraySet`)
- Rewrite of Json reading/writing using Kotlinx Serialization library
- Updates minsdk from 17 to 23

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,14 @@
 ## Change Log
 
+### Unreleased
+- Upgraded several dependencies
+- Upgraded to AGP 8.2.1
+- Upgraded to Kotlin 1.9.22
```
Comment (sklinefelter) [inline CHANGELOG.md:6]: Concern here - This would be a breaking change I'm pretty sure since bumping AGP and Kotlin would require the client to have the same Kotlin metadata (baked into AGP).  The app I'm working on currently is not Kotlin 1.9.22 ready so it would take some effort for us to pick this version of Harmony up.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,14 @@
...
+- Upgraded to Gradle 8.5
+- MinSDK is now 23
+- Rewrite of JSON reading/writing
+- Use library version rather than OS APIs (ex. `androidx.collection.ArraySet` vs `android.util.ArraySet`)
```
Comment (sklinefelter) [inline CHANGELOG.md:10]: 👍🏽 This seems like a good thing

File: buildSrc/src/main/kotlin/com/frybits/harmony/gradle/FrybitsApplicationPlugin.kt [hunk 3]
```diff
@@ -33,10 +33,10 @@ import org.jetbrains.kotlin.gradle.plugin.KotlinAndroidPluginWrapper
 
 class FrybitsApplicationPlugin: Plugin<Project> {
 
-    override fun apply(target: Project) {
-        target.applyAppPlugins()
+    override fun apply(target: Project) = target.run {
```
Comment (sklinefelter) [inline buildSrc/src/main/kotlin/com/frybits/harmony/gradle/FrybitsApplicationPlugin.kt:36]: Question - This shouldn't be any functional change, correct?

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,14 @@
 ## Change Log
 
+### Unreleased
+- Upgraded several dependencies
+- Upgraded to AGP 8.2.1
+- Upgraded to Kotlin 1.9.22
+- Upgraded to Gradle 8.5
+- MinSDK is now 23
```
Comment (sklinefelter) [inline CHANGELOG.md:8]: Breaking change, however the minSDK version we're on is SDK 26 so we should be able to pick this one up relatively easy.

File: gradle/libs.versions.toml [hunk 5]
```diff
@@ -1,32 +1,33 @@
 [versions]
 # Kotlin
-kotlin = "1.9.10"
+kotlin = "1.9.22"
```
Comment (sklinefelter) [inline gradle/libs.versions.toml:3]: 🤔 We were on 1.2.2 of harmony before, so maybe my earlier comment about Kotlin update is less concerning.  I know we had a heck of a time with getting to Kotlin 1.8.x and AGP 8

File: harmony/src/main/java/com/frybits/harmony/internal/HarmonyJsonUtils.kt [hunk 6]
```diff
@@ -46,141 +61,123 @@ private const val BOOLEAN = "boolean"
...
 private const val SET = "set"
 
+private val json = Json {
+    prettyPrint = false
+    isLenient = false
+}
+
+@JvmSynthetic
+@Throws(IOException::class, SerializationException::class)
+internal fun <T : Reader> T.readHarmony(): Pair<String?, Map<String?, Any?>> {
```
Comment (sklinefelter) [inline harmony/src/main/java/com/frybits/harmony/internal/HarmonyJsonUtils.kt:71]: Crazy thought/question - Could we configure from the client side which JSON serializer to use?  Thinking about it like this:

1. We take in the new version of harmony 
2. Configure the Harmony SDK to have x% of users leveraging the new json serializer and rest still chugging along on the old serializer

I think that makes it easier for us to safely adopt and migrate and we could report issues that we see as we're onboarding.  We could also just force the old serializer if we hit something big enough that causes trouble.

File: harmony/src/main/java/com/frybits/harmony/Harmony.kt [hunk 7]
```diff
@@ -285,12 +285,12 @@ private class HarmonyImpl constructor(
...
             _InternalHarmonyLog.e(LOG_TAG, "IllegalStateException while reading data file", e)
             null to emptyMap()
-        } catch (e: JSONException) {
-            _InternalHarmonyLog.e(LOG_TAG, "JSONException while reading data file", e)
-            null to emptyMap()
         } catch (e: IOException) {
             _InternalHarmonyLog.e(LOG_TAG, "IOException occurred while reading json", e)
             null to emptyMap()
+        } catch (e: SerializationException) {
+            _InternalHarmonyLog.e(LOG_TAG, "Unable to parse json string", e)
```
Comment (sklinefelter) [inline harmony/src/main/java/com/frybits/harmony/Harmony.kt:292]: In this case would the client receive the default value?

File: harmony/src/androidTest/java/com/frybits/harmony/test/HarmonyJsonUtilsTest.kt [hunk 8]
```diff
@@ -73,6 +73,11 @@ private const val MIXED_JSON = """
                 "type": "int",
                 "key": "count",
                 "value": 3
+            },
+            {
+                "type": "set",
```
Comment (sklinefelter) [inline harmony/src/androidTest/java/com/frybits/harmony/test/HarmonyJsonUtilsTest.kt:78]: yay tests

File: gradle/libs.versions.toml [hunk 9]
```diff
@@ -1,32 +1,33 @@
...
+androidx-navigation = "2.7.6"
 androidx-appcompat = "1.6.1"
 androidx-core = "1.12.0"
-androidx-test-core = "1.5.0"
-androidx-lifecycle = "2.6.2"
+androidx-lifecycle = "2.7.0"
 androidx-constraintlayout = "2.1.4"
-androidx-room = "2.5.2"
-androidx-annotation = "1.7.0"
+androidx-room = "2.6.1"
```
Comment (sklinefelter) [inline gradle/libs.versions.toml:17]: Suggestion - Maybe the version bumps should be in a different PR?  Seems like they were all shotgun updated which would make it a little problematic to isolate a particular version update or set of updates that lead to any misbehavior.

File: buildSrc/src/main/kotlin/com/frybits/harmony/gradle/FrybitsApplicationPlugin.kt (hunk 3 above)
Comment (pablobaxter) [reply buildSrc/src/main/kotlin/com/frybits/harmony/gradle/FrybitsApplicationPlugin.kt:36]: No, this a minor change for the build logic and is not part of Harmony.

File: CHANGELOG.md (hunk 4 above)
Comment (pablobaxter) [reply CHANGELOG.md:8]: MinSDK version was updated due to build issues with tests. It's not absolutely necessary, but I didn't want to deal with the headache of supporting Android 4.x.

File: gradle/libs.versions.toml (hunk 5 above)
Comment (pablobaxter) [reply gradle/libs.versions.toml:3]: Upgrading these plugins aren't an issue for consumers of Harmony, unless you are compiling from source.

File: harmony/src/main/java/com/frybits/harmony/Harmony.kt (hunk 7 above)
Comment (pablobaxter) [reply harmony/src/main/java/com/frybits/harmony/Harmony.kt:292]: This function reads the preferences from the `prefs.data` file, which holds the current state of all preferences. If there is a failure in reading the file, this is equivalent to clearing preferences. Similar logic exists for [`SharedPreferences`](https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/core/java/android/app/SharedPreferencesImpl.java;l=207?q=SharedPreferencesImp).

File: gradle/libs.versions.toml (hunk 9 above)
Comment (pablobaxter) [reply gradle/libs.versions.toml:17]: Many of these updates only affect the demo/test app for Harmony. The only new libraries included for Harmony are the update to the annotations, and the Kotlin Serialization library.

**Review by sklinefelter — COMMENTED** [review]
I think my two biggest concerns are:

- Kotlin/AGP bump
- Me not knowing enough about the implementation and new serialization library to ask many questions about "gotchas"

Just to clarify the goal here, the new JSON serializer should resolve the issue or at the very least give a better stacktrace or indication as to what is going on, yeah?

That being said, thank you!  I don't have objections to trying these changes out.  Any option for a SNAPSHOT or pre-release (I know I am asking for a lot 😅)

File: harmony/src/main/java/com/frybits/harmony/internal/HarmonyJsonUtils.kt [hunk 10]
```diff
@@ -46,141 +61,123 @@ private const val BOOLEAN = "boolean"
...
+        putJsonObject(METADATA) {
+            put(NAME_KEY, prefsName)
+        }
+
+        putJsonArray(DATA) {
+            data.forEach { (k, v) ->
+                // Only add a json object if expected type is found. Ignore all other types.
+                when (v) {
+                    is Int -> {
+                        addJsonObject {
```
[the remark this answers is not in the document]
Comment (pablobaxter) [reply harmony/src/main/java/com/frybits/harmony/internal/HarmonyJsonUtils.kt:127]: Yes. The JSON file is the state of the preferences, with a transaction file that writes the interim changes coming from both processes. When this transaction file hits a limit or a new process is started, it's flushed, and the JSON is re-updated.

File: harmony/src/main/java/com/frybits/harmony/internal/HarmonyJsonUtils.kt [hunk 11]
```diff
@@ -46,141 +61,123 @@ private const val BOOLEAN = "boolean"
...
-                    if (currName == METADATA) {
-                        beginObject()
-                        val n = nextName()
-                        if (n == NAME_KEY) {
-                            prefsName = nextString()
+internal fun <T : Writer> T.putHarmony(prefsName: String, data: Map<String?, Any?>): T {
+    val json = buildJsonObject {
+
+        // Still not used, but left for future use
+        putJsonObject(METADATA) {
```
[the remark this answers is not in the document]
Comment (pablobaxter) [reply harmony/src/main/java/com/frybits/harmony/internal/HarmonyJsonUtils.kt:118]: It's in case the JSON format ever changes. It allows a way to version the serialization.

**Comment by pablobaxter** [comment]
> I think my two biggest concerns are:
> 
>     * Kotlin/AGP bump
> 
>     * Me not knowing enough about the implementation and new serialization library to ask many questions about "gotchas"
> 
> 

> 


> Just to clarify the goal here, the new JSON serializer should resolve the issue or at the very least give a better stacktrace or indication as to what is going on, yeah?

Yes

> Any option for a SNAPSHOT or pre-release

I can get something published for this later today. If you want to play with this sooner, you can just publish locally with 
```
./gradlew publishToMavenLocal -PRELEASE_SIGNING_ENABLED=false
```
